File: app/src/main/java/me/raatiniemi/worker/data/service/ongoing/ResumeService.java
/*
 * Copyright (C) 2017 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.data.service.ongoing;

import android.content.Intent;

import java.util.Date;

import me.raatiniemi.worker.R;
import me.raatiniemi.worker.domain.exception.ActiveProjectException;
import me.raatiniemi.worker.domain.interactor.ClockIn;
import me.raatiniemi.worker.domain.interactor.GetProject;
import me.raatiniemi.worker.domain.model.Project;
import me.raatiniemi.worker.features.shared.view.notification.ErrorNotification;
import me.raatiniemi.worker.features.shared.view.notification.PauseNotification;
import timber.log.Timber;

public class ResumeService extends OngoingService {
    public ResumeService() {
        super("ResumeService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        long projectId = getProjectId(intent);

        try {
            clockInProjectNow(projectId);

            updateUserInterface(projectId);

            if (isOngoingNotificationEnabled()) {
                GetProject getProject = buildGetProjectUseCase();
                Project project = getProject.execute(projectId);

                sendPauseNotification(project);
                return;
            }

            dismissResumeNotification(projectId);
        } catch (Exception e) {
            Timber.w(e, "Unable to resume project");

            sendErrorNotification(projectId);
        }
    }

    private void clockInProjectNow(long projectId) {
        try {
            ClockIn clockIn = buildClockInUseCase();
            clockIn.execute(projectId, new Date());
        } catch (ActiveProjectException e) {
            Timber.e(e, "Resume service called with active project");
        }
    }

    ClockIn buildClockInUseCase() {
        return new ClockIn(getTimeIntervalRepository());
    }

    GetProject buildGetProjectUseCase() {
        return new GetProject(getProjectRepository());
    }

    private void sendPauseNotification(Project project) {
        sendNotification(
                project.getId(),
                PauseNotification.build(this, project, isOngoingNotificationChronometerEnabled())
        );
    }

    private void dismissResumeNotification(long projectId) {
        dismissNotification(projectId);
    }

    private void sendErrorNotification(long projectId) {
        sendNotification(
                projectId,
                ErrorNotification.buildOngoing(
                        this,
                        getString(R.string.error_notification_resume_title),
                        getString(R.string.error_notification_resume_message)
                )
        );
    }
}

File: app/src/main/java/me/raatiniemi/worker/data/service/ongoing/ResumeService.kt
/*
 * Copyright (C) 2018 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.data.service.ongoing

import android.content.Intent
import me.raatiniemi.worker.domain.exception.ActiveProjectException
import me.raatiniemi.worker.domain.model.Project
import me.raatiniemi.worker.domain.usecase.CalculateTimeToday
import me.raatiniemi.worker.domain.usecase.ClockIn
import me.raatiniemi.worker.domain.usecase.GetProject
import me.raatiniemi.worker.features.shared.view.notification.PauseNotification
import me.raatiniemi.worker.monitor.analytics.Event
import me.raatiniemi.worker.monitor.analytics.UsageAnalytics
import org.koin.android.ext.android.inject
import timber.log.Timber
import java.util.*

internal class ResumeService : OngoingService("ResumeService") {
    private val usageAnalytics: UsageAnalytics by inject()
    private val clockIn: ClockIn by inject()
    private val getProject: GetProject by inject()
    private val calculateTimeToday: CalculateTimeToday by inject()

    override fun onHandleIntent(intent: Intent?) {
        try {
            val projectId = getProjectId(intent)
            val project = getProject(projectId)

            clockIn(project)

            updateUserInterface(project.id.value)
            sendOrDismissPauseNotification(project)
        } catch (e: Exception) {
            Timber.e(e, "Unable to resume project")
        }
    }

    private fun clockIn(project: Project) {
        try {
            clockIn(project.id.value, Date())

            usageAnalytics.log(Event.NotificationClockIn)
        } catch (e: ActiveProjectException) {
            Timber.w(e, "Resume service called with active project")
        }
    }

    private fun sendOrDismissPauseNotification(project: Project) {
        sendOrDismissOngoingNotification(project) {
            PauseNotification.build(
                this,
                project,
                calculateTimeToday(project),
                isOngoingNotificationChronometerEnabled
            )
        }
    }
}
